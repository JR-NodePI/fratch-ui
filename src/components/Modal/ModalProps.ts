import { ReactNode } from 'react';

export const ModalTypes = {
  ACCEPT: 'accept' as const,
  CONFIRM: 'confirm' as const,
  INFO: 'info' as const,
};

export const ModalCloseTypes = {
  ACCEPT: 'accept' as const,
  CANCEL: 'cancel' as const,
  CLOSE: 'close' as const,
};

export type ModalCloseType = (typeof ModalCloseTypes)[keyof typeof ModalCloseTypes];

export type ModalProps = {
  acceptButtonLabel?: string;
  cancelButtonLabel?: string;
  children: ReactNode;
  onClose?: (type: ModalCloseType) => void;
  onOpen?: () => void;
  title?: ReactNode;
  type?: (typeof ModalTypes)[keyof typeof ModalTypes];
  visible?: boolean;
};
