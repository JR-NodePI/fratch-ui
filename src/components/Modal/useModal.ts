import { useContext } from 'react';

import ModalContext from './ModalContext';
import type { ModalContextProps } from './ModalProps';

export default function useModal(): ModalContextProps {
  return useContext(ModalContext);
}
