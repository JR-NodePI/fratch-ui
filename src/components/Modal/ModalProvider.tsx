import { ReactNode, useCallback, useMemo, useState } from 'react';

import Modal from './Modal';
import { ModalTypes } from './ModalConstants';
import ModalContext from './ModalContext';
import {
  ModalCloseType,
  ModalContextProps,
  ModalProps,
  ModalType,
  ShowModalProps,
} from './ModalProps';

const defaultModalProps = {
  children: <></>,
  title: '',
  visible: false,
};

function ModalProvider({ children }: { children: ReactNode }): JSX.Element {
  const [modalProps, setModalProps] = useState<ModalProps>(defaultModalProps);

  const handleModalOpen = useCallback(
    (props: ShowModalProps, type: ModalType) => {
      setModalProps(state => {
        if (state.visible) {
          return state;
        }

        const handleClose = (type: ModalCloseType): void => {
          setModalProps(defaultModalProps);
          props.onClose?.(type);
        };

        return {
          ...props,
          children: props.content,
          onClose: handleClose,
          type,
          visible: !state.visible,
        };
      });
    },
    []
  );

  const value = useMemo<ModalContextProps>(
    () => ({
      showModalAccept: (props: ShowModalProps): void => {
        handleModalOpen(props, ModalTypes.ACCEPT);
      },
      showModalConfirm: (props): void => {
        handleModalOpen(props, ModalTypes.CONFIRM);
      },
      showModalInfo: (props): void => {
        handleModalOpen(props, ModalTypes.INFO);
      },
    }),
    [handleModalOpen]
  );

  return (
    <ModalContext.Provider value={value}>
      {children}
      <Modal {...modalProps} />
    </ModalContext.Provider>
  );
}

export default ModalProvider;
