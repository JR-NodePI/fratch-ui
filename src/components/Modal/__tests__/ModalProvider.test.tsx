import { act } from 'react-dom/test-utils';

import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { debounce } from 'lodash';
import { describe, expect, it, vi } from 'vitest';

import {
  MODAL_TIMEOUT_TO_CLOSE,
  MODAL_TIMEOUT_TO_OPEN,
} from '../ModalConstants';
import { type ShowModalProps } from '../ModalProps';
import ModalProvider from '../ModalProvider';
import useModal from '../useModal';

vi.mock('lodash', async requireActual => {
  const module = (await requireActual()) as { debounce: typeof debounce };
  return { debounce: vi.fn(fn => module.debounce(fn, 10)) };
});

describe('Modal', () => {
  const modalId = 'crypto-mock-random-UUID-1234';

  let triggerShowModalAccept: (props: ShowModalProps) => void;
  let triggerShowModalConfirm: (props: ShowModalProps) => void;
  let triggerShowModalInfo: (props: ShowModalProps) => void;

  const MockComponent = (): JSX.Element => {
    const { showModalAccept, showModalConfirm, showModalInfo } = useModal();

    triggerShowModalAccept = showModalAccept;
    triggerShowModalConfirm = showModalConfirm;
    triggerShowModalInfo = showModalInfo;

    return <p>Mock contents</p>;
  };

  const getComponent = (): JSX.Element => (
    <ModalProvider>
      <MockComponent />
    </ModalProvider>
  );

  const setup = (): ReturnType<typeof render> => render(getComponent());

  beforeEach(() => {
    vi.spyOn(crypto, 'randomUUID').mockReturnValue(modalId);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('should render properly', () => {
    const { baseElement } = setup();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render an accept Modal', () => {
    const { baseElement } = setup();

    act(() => {
      triggerShowModalAccept({
        title: 'Accept modal title',
        content: <>Accept modal content</>,
      });
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render a confirm Modal', () => {
    const { baseElement } = setup();

    act(() => {
      triggerShowModalConfirm({
        title: 'Confirm modal title',
        content: <>Confirm modal content</>,
      });
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render a info Modal', () => {
    const { baseElement } = setup();

    act(() => {
      triggerShowModalInfo({
        title: 'Info modal title',
        content: <>Info modal content</>,
      });
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should call onOpen and onClose an accept Modal', async () => {
    setup();

    const onOpen = vi.fn();
    const onClose = vi.fn();

    act(() => {
      triggerShowModalAccept({
        title: 'Accept modal title',
        content: <>Accept modal content</>,
        onClose,
        onOpen,
      });
    });

    expect(onOpen).toHaveBeenCalledTimes(1);
    expect(onOpen).toHaveBeenCalledWith();

    await userEvent.click(screen.getByRole('button', { name: /Close/i }));

    await waitFor(() => {
      expect(onClose).toHaveBeenCalledTimes(1);
      expect(onClose).toHaveBeenCalledWith('close');
    });

    expect(debounce).toHaveBeenCalledTimes(2);
    expect(debounce).toHaveBeenNthCalledWith(
      1,
      expect.any(Function),
      MODAL_TIMEOUT_TO_OPEN
    );
    expect(debounce).toHaveBeenNthCalledWith(
      2,
      expect.any(Function),
      MODAL_TIMEOUT_TO_CLOSE
    );
  });
});
