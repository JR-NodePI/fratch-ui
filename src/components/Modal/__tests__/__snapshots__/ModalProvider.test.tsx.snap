// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > should render a confirm Modal 1`] = `
<body>
  <div>
    <p>
      Mock contents
    </p>
  </div>
  <div
    aria-label=""
    class="_modal_overflow_f57fcc"
  >
    <section
      class="_modal_f57fcc _confirm_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <header
        class="_header_f57fcc"
      >
        <h5
          class="_title_f57fcc"
        >
          Confirm modal title
        </h5>
      </header>
      <div
        class="_content_f57fcc"
      >
        Confirm modal content
      </div>
      <footer
        class="_footer_f57fcc"
      >
        <button
          class="_button_2a1efe _default_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            Cancel
          </span>
        </button>
        <button
          class="_button_2a1efe _tertiary_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            OK
          </span>
        </button>
      </footer>
    </section>
  </div>
</body>
`;

exports[`Modal > should render a info Modal 1`] = `
<body>
  <div>
    <p>
      Mock contents
    </p>
  </div>
  <div
    aria-label="Close modal"
    class="_modal_overflow_f57fcc"
  >
    <section
      class="_modal_f57fcc _info_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <header
        class="_header_f57fcc"
      >
        <h5
          class="_title_f57fcc"
        >
          Info modal title
        </h5>
      </header>
      <button
        class="_button_closer_8d9a3f _button_closer_f57fcc"
        title="Close"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _icon_8d9a3f"
          fill="none"
          viewBox="-1 -1 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Close
        </i>
      </button>
      <div
        class="_content_f57fcc"
      >
        Info modal content
      </div>
    </section>
  </div>
</body>
`;

exports[`Modal > should render an accept Modal 1`] = `
<body>
  <div>
    <p>
      Mock contents
    </p>
  </div>
  <div
    aria-label="Close modal"
    class="_modal_overflow_f57fcc"
  >
    <section
      class="_modal_f57fcc _accept_f57fcc"
      id="crypto-mock-random-UUID-1234"
    >
      <header
        class="_header_f57fcc"
      >
        <h5
          class="_title_f57fcc"
        >
          Accept modal title
        </h5>
      </header>
      <button
        class="_button_closer_8d9a3f _button_closer_f57fcc"
        title="Close"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _icon_8d9a3f"
          fill="none"
          viewBox="-1 -1 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Close
        </i>
      </button>
      <div
        class="_content_f57fcc"
      >
        Accept modal content
      </div>
      <footer
        class="_footer_f57fcc"
      >
        <button
          class="_button_2a1efe _primary_2a1efe _small_2a1efe _button_f57fcc"
        >
          <span>
            Accept
          </span>
        </button>
      </footer>
    </section>
  </div>
</body>
`;

exports[`Modal > should render properly 1`] = `
<body>
  <div>
    <p>
      Mock contents
    </p>
  </div>
</body>
`;
